### DISCUSSION TITLE
자료구조 - 배열/연결리스트 - 은진&연서

### DISCUSSION THREAD

**Discussion by aeroej** [Shortudy-10th/tech-for-developer#40] [질의 응답]
- 배열과 연결리스트의 특징에 대해서 설명해주세요.
- Static Array와 Dynamic Array의 차이점은 무엇일까요?
- Array는 Random Access Data Structure의 하나의 예가 될 수 있다고 합니다. Random Access란 무엇일까요?
- 연결리스트의 종류에 대해 설명해주세요.
- 배열과 연결리스트의 특성에 기반하여, 각 자료구조를 사용하는 것이 다른 자료구조를 사용하는 것보다 유리한 상황에 대해 설명해주세요.
- 연결리스트에서 요소를 추가하거나 삭제하는 과정을 설명해주세요.

**Comment by aeroej** [comment]
### ❓️배열과 연결리스트의 특징에 대해서 설명해주세요.

<br>

### 💬 답변
<blockquote>

**공통점** 
순차적으로 데이터를 저장, 탐색할 수 있다.

**차이점**
1. 메모리 구조
- 배열 : 메모리의 연속적인 공간에 데이터를 저장한다. -> 빠른 탐색이 가능하다.
- 연결 리스트 : 메모리의 비연속적인 공간에 데이터를 저장한다. 한 요소에 value와 다음 요소의 주소값이 포함된다.

2. 크기
- 배열 : 크기 고정 -> 크기를 변경하려면 새 배열을 생성해야 한다.
- 연결 리스트 : 동적으로 크기 변경 -> 요소의 추가 및 삭제가 용이하다.

3. 시간복잡도
- 배열 : O(1) -> 인덱스 덕분!
- 연결 리스트 : O(N) -> 순차적으로 탐색해야 하기 때문!



</blockquote>

<br>

**Comment by aeroej** [comment]
### ❓️Static Array와 Dynamic Array의 차이점은 무엇일까요?

<br>

### 💬 답변

> 1. 크기
 Static Array: 고정된 크기
 Dynamic Array: 가변적 크기
>
> 2. 메모리
Static Array는 Stack영역에 컴파일 시 메모리 할당 
Dynamic Array는 실행시간에 Heap영역에 메모리 할당


<br>

**Comment by aeroej** [comment]
### ❓️Array는 Random Access Data Structure의 하나의 예가 될 수 있다고 합니다. Random Access란 무엇일까요?

<br>

### 💬 답변
> 
> ![image](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/215e9939-78c5-4fa7-8798-8bd9d5b79482)
> 데이터 구조 내에서 특정 위치에 있는 요소에 직접적으로 접근하는 능력을 말한다.
> 배열의 경우, arr[1]에 접근하든 arr[2]에 접근하든 접근 시간은 똑같이 상수시간 O(1)이다.

<br>

**Reply by watty-exclamationMark** [reply]
이거 배열의 특징 중 하나라 합쳐도 좋을것같을지도..

**Comment by aeroej** [comment]
### ❓️ 연결리스트의 종류에 대해 설명해주세요.

<br>

### 💬 답변
> 1. 단일 연결 리스트 
>![image](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/2bc69d9c-a75c-4fc8-858d-c76ecb4b6ca9)
> 가장 단순한 구현으로, 각 노드는 그 다음 노드에 대한 참조를 가지고 있는 단방향 연결 리스트이다.
>
> 2. 이중 연결 리스트
> ![image](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/2377d144-e115-49b4-a208-956bd4a304e7)
> 단일 연결 리스트와는 다르게 이전 노드의 참조까지 갖고 있는 양뱡향 연결 리스트이다.
> 
> 3. 원형 연결 리스트
> ![image](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/d0dcc42b-a7b3-4349-982f-64e86b4665ce)
> 가장 마지막 노드를 가장 첫 번째 노드를 가리키게 해 원형 구조를 만든 리스트이다.

<br>

**Reply by watty-exclamationMark** [reply]
각각 어떤 상황에서 사용해야 좋을까요?

**Comment by aeroej** [comment]
### ❓️ 배열과 연결리스트의 특성에 기반하여, 각 자료구조를 사용하는 것이 다른 자료구조를 사용하는 것보다 유리한 상황에 대해 설명해주세요.

<br>

### 💬 답변
> 1. 배열
> + 빠른 랜덤 엑세스 
>  + 연속된 메모리 공간 사용
> + 보통 크기가 고정됨
>
> -> 이를 고려해볼 때 배열은 인덱스에 빠르게 접근해야 하는 경우, 메모리 효율성이 중요한 경우에 사용하는 것이 좋다.
>
> 2. 연결리스트
> + 동적으로 관리 가능
> + 메모리 할당의 유연성
> + 삽입 삭제가 배열보다 빠르다
>
> -> 이를 고려해볼 때 연결리스트는 삽입, 삭제가 많이 일어날 경우, 유연한 메모리 할당이 필요할 때 사용하는 것이 좋다.

<br>

**Reply by watty-exclamationMark** [reply]
이것두,, 특징이랑 합치면 되지 않을까 싶기도

**Comment by aeroej** [comment]
### ❓️ 연결리스트에서 요소를 추가하거나 삭제하는 과정을 설명해주세요.

<br>

### 💬 답변
> 1. 추가하는 경우
> 추가가 필요한 인덱스에 접근, 추가 노드가 가리키는 다음 노드를 접근한 노드가 참조하고 있는 다음 노드를 가리키게 한다.
> 그 뒤 본래 접근한 인덱스의 노드가 가리키는 다음 노드를 추가한 노드로 설정한다.
> 
> ![image](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/70e9e194-37b7-4695-9751-2448d09e7e2c)
>
>
>2. 삭제하는 경우
> 삭제 대상의 이전 노드로 가서, 이전 노드가 가리키는 다음 노드를 삭제 대상의 다음 노드로 바꾼다.
> 
>![image](https://github.com/Shortudy-10th/tech-for-developer/assets/70586307/db40ab01-46d7-4ac9-8a7d-9e13234d5705)

<br>
